pkg/protocol/engine/mempool/v1/mempool.go

```diff
@@ -273,12 +280,22 @@ func (m *MemPool[VotePower]) requestStateWithMetadata(stateReference ledger.Stat
 }
 
 func (m *MemPool[VotePower]) forkTransaction(transaction *TransactionMetadata, input *StateMetadata) {
-	switch err := m.conflictDAG.CreateConflict(transaction.ID(), transaction.conflictIDs.Get(), advancedset.New(input.ID()), acceptance.Pending); {
+	switch err := m.conflictDAG.CreateConflict(transaction.ID(), advancedset.New(input.ID()), acceptance.Pending); {
 	case errors.Is(err, conflictdag.ErrConflictExists):
-		m.conflictDAG.JoinConflictSets(transaction.ID(), advancedset.New(input.ID()))
+		err = m.conflictDAG.JoinConflictSets(transaction.ID(), advancedset.New(input.ID()))
+		if err != nil {
+			panic(err)
+		}
```

Comment:
Instead of iterating over all inputs in `bookTransaction` I think it would be better to just create the conflict right away with all its inputs. Creating it in multiple rounds and then joining it together seems to be prone to race conditions or other weird side effects. Also just makes it more computation heavy for no reason.

Apart from that I think the API is not great if it returns an error and it is actually part of the normal programming flow. Would be better to have an `CreateOrUpdate` conflict method.

Reply:
Good ideas! Already applied them!